### PR TITLE
chore: refactor code to decrease complexity

### LINKED ISSUE #533: Refactor code and increase test coverage
Pre-requisites:

- [x] https://github.com/open-feature/open-feature-operator/issues/529
- [x] https://github.com/open-feature/flagd/issues/983
- [x] https://github.com/open-feature/open-feature-operator/issues/531

## Goal

Clean up logic of packages in OFO

### Details

There are multiple possibilities to improve the maintability of OFO logic, especially we are talking about the following:

- [ ] polish controller logic; increase unit tests coverage
- [x] polish flagdproxy handler
- [x] polish flagdinjector
- [x] create global `common` package and move `controllers/common` into it
- [x] move `pkg` into global `common` package
- [x] polish and increase coverage of `common` package
- [x] polish and increase coverage of `webhook` package -> some functions can be moved to a local `webhooks/common` package for better readibility
- [x] remove all nolint comments (where it makes sense)

### DoD

- [x] code is better structured and easier to read/understand
- [x] test coverage is increased in webhook and common packages

### Notes

Some of the adjustments can be found in the [PoC](https://github.com/open-feature/open-feature-operator/pull/527)

### PR DESCRIPTION
**Pull request by odubajDT**
Fixes #533

## Changes

- refactor code to remove `nolint` comments (where it makes sense)
- move `common/constants` package to `common` right next to the rest of constants
- increase test coverage for `common` package
- increase test coverage for `FlagdProxyHandler`
- polish `flagdinjector`

### CODE REVIEW

File: common/flagdinjector/flagdinjector.go [hunk 1]
```diff
@@ -195,40 +193,49 @@ func (fi *FlagdContainerInjector) buildSources(ctx context.Context, objectMeta *
...
+func (fi *FlagdContainerInjector) createSourceConfig(ctx context.Context, source api.Source, objectMeta *metav1.ObjectMeta, podSpec *corev1.PodSpec, sidecar *corev1.Container) (*types.SourceConfig, error) {
+	var sourceCfg types.SourceConfig
+	var err error
+
+	switch {
+	case source.Provider.IsKubernetes():
+		sourceCfg, err = fi.toKubernetesProviderConfig(ctx, objectMeta, podSpec, source)
+		if err != nil {
+			return nil, err
+		}
```
Comment (thisthat) [inline common/flagdinjector/flagdinjector.go:217]: we can remove these returns of the errors and change the last one to be `return &sourceCfg, err`

**Review by Kavindu-Dodan — APPROVED** [review]

File: common/common.go [hunk 2]
```diff
@@ -2,46 +2,38 @@ package common
...
-	FinalizerName               = "featureflag.core.openfeature.dev/finalizer"
-	OpenFeatureAnnotationPath   = "spec.template.metadata.annotations.openfeature.dev/openfeature.dev"
-	FeatureFlagSourceAnnotation = "featureflagsource"
-	OpenFeatureAnnotationRoot   = "openfeature.dev"
+	ReconcileErrorInterval                          = 10 * time.Second
+	ReconcileSuccessInterval                        = 120 * time.Second
+	FinalizerName                                   = "featureflag.core.openfeature.dev/finalizer"
+	OpenFeatureAnnotationPath                       = "spec.template.metadata.annotations.openfeature.dev/openfeature.dev"
+	OpenFeatureAnnotationRoot                       = "openfeature.dev"
+	FlagDImagePullPolicy          corev1.PullPolicy = "Always"
```
Comment (beeme1mr) [inline common/common.go:22]:
```suggestion
	FlagdImagePullPolicy          corev1.PullPolicy = "Always"
```

File: main.go [hunk 3]
```diff
@@ -213,21 +189,12 @@ func main() {
...
-					corev1.ResourceCPU:    cpuRequestResource,
-					corev1.ResourceMemory: ramRequestResource,
-				},
-			},
-			Image: env.SidecarImage,
-			Tag:   env.SidecarTag,
+			Client:                    mgr.GetClient(),
+			Logger:                    ctrl.Log.WithName("flagd-container injector"),
+			FlagdProxyConfig:          kph.Config(),
+			FlagDResourceRequirements: *resources,
```
Comment (beeme1mr) [inline main.go:195]:
```suggestion
			FlagdResourceRequirements: *resources,
```

File: common/flagdinjector/flagdinjector.go [hunk 4]
```diff
@@ -381,7 +379,7 @@ func (fi *FlagdContainerInjector) generateBasicFlagdContainer(flagSourceConfig *
 		Args: []string{
 			"start",
 		},
-		ImagePullPolicy: constant.FlagDImagePullPolicy,
+		ImagePullPolicy: common.FlagDImagePullPolicy,
```
Comment (beeme1mr) [inline common/flagdinjector/flagdinjector.go:382]:
```suggestion
		ImagePullPolicy: common.FlagdImagePullPolicy,
```

File: main.go (hunk 3 above)
Comment (beeme1mr) [reply main.go:195]: It doesn't look like this is used in any of the changed files. Please update all the references to this.

**Review by beeme1mr — COMMENTED** [review]
There are a few variables that use improper casing of flagd. It would be nice to address them either in this PR or in a follow up.

File: controllers/core/featureflagsource/controller.go [hunk 5]
```diff
@@ -44,6 +44,9 @@ type FeatureFlagSourceReconciler struct {
 	FlagdProxy *flagdproxy.FlagdProxyHandler
 }
 
+// renovate: datasource=github-tags depName=open-feature/flagd/flagd-proxy
+const flagdProxyTag = "v0.3.0"
```
Comment (bacherfl) [inline controllers/core/featureflagsource/controller.go:48]: i guess this can already be set to `v0.7.0` here

File: common/types/envconfig.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	FlagdProxyTag            string `envconfig:"FLAGD_PROXY_TAG" default:"v0.3.0"`
+	FlagdProxyPort           int    `envconfig:"FLAGD_PROXY_PORT" default:"8015"`
+	FlagdProxyManagementPort int    `envconfig:"FLAGD_PROXY_MANAGEMENT_PORT" default:"8016"`
+	FlagdProxyDebugLogging   bool   `envconfig:"FLAGD_PROXY_DEBUG_LOGGING" default:"false"`
+
+	SidecarEnvVarPrefix   string `envconfig:"SIDECAR_ENV_VAR_PREFIX" default:"FLAGD"`
+	SidecarManagementPort int    `envconfig:"SIDECAR_MANAGEMENT_PORT" default:"8014"`
+	SidecarPort           int    `envconfig:"SIDECAR_PORT" default:"8013"`
+	SidecarImage          string `envconfig:"SIDECAR_IMAGE" default:"ghcr.io/open-feature/flagd"`
+	// renovate: datasource=github-tags depName=open-feature/flagd/flagd-proxy
```
Comment (bacherfl) [inline common/types/envconfig.go:16]: I think the sidecar should refer to flagd, not flagd-proxy

```suggestion
	// renovate: datasource=github-tags depName=open-feature/flagd/flagd
```

File: controllers/core/featureflagsource/controller.go (hunk 5 above)
Comment (bacherfl) [reply controllers/core/featureflagsource/controller.go:48]: just checked, and the latest proxy release seems to be 0.3.0, so please disregard my comment

Comment (odubajDT) [reply controllers/core/featureflagsource/controller.go:48]: This is flagd-proxy, its version is just 0.3.0, see here https://github.com/open-feature/flagd/tags

File: common/types/envconfig.go (hunk 6 above)
Comment (odubajDT) [reply common/types/envconfig.go:16]: good catch!

**Review by bacherfl — APPROVED** [review]

**Pull request merged**